`python3Packages.pycurl` building failed

### Describe the bug
A clear and concise description of what the bug is.
Failed to run `nixos-rebuild boot`, caused by the failure of building `pycurl`.
Run `nix-shell -p python3Packages.pycurl` confirms this proflem


### Steps To Reproduce
Steps to reproduce the behavior:
1. run ``nix-shell -p python3Packages.pycurl`
2. fails

### Expected behavior
A clear and concise description of what you expected to happen.
`nix-shell -p python3Packages.pycurl` should succeed.

### Screenshots
If applicable, add screenshots to help explain your problem.

### Additional context
Add any other context about the problem here.

### Notify maintainers
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->


### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.30, NixOS, 22.05 (Quokka), 22.05pre364094.12417777b22`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

`pycurl-7.43.0.6` builds fine (using nix-shell)

`pycurl-7.45.1` on unstable fails with:
```
builder for '/nix/store/z95037pm13nbfmqnixy1npgbap1a1zs9-python3.9-pycurl-7.45.1.drv' failed with exit code 1;
       last 10 log lines:
       > test_post_with_null_byte passed 1 out of the required 1 times. Success!
       >
       > ===End Flaky Test Report===
       > =========================== short test summary info ============================
       > FAILED tests/cadata_test.py::CaCertsTest::test_request_with_verifypeer - pycu...
       > FAILED tests/certinfo_test.py::CertinfoTest::test_getinfo_raw_certinfo - pycu...
       > FAILED tests/certinfo_test.py::CertinfoTest::test_request_with_certinfo - pyc...
       > FAILED tests/certinfo_test.py::CertinfoTest::test_request_without_certinfo - ...
```

--- a/pkgs/tools/networking/curl/7.82.0-openssl-fix-CN-check.patch
+++ b/pkgs/tools/networking/curl/7.82.0-openssl-fix-CN-check.patch
@@ -1,0 +1,30 @@
+From 911714d617c106ed5d553bf003e34ec94ab6a136 Mon Sep 17 00:00:00 2001
+From: Daniel Stenberg <daniel@haxx.se>
+Date: Tue, 8 Mar 2022 13:38:13 +0100
+Subject: [PATCH] openssl: fix CN check error code
+
+Due to a missing 'else' this returns error too easily.
+
+Regressed in: d15692ebb
+
+Reported-by: Kristoffer Gleditsch
+Fixes #8559
+Closes #8560
+---
+ lib/vtls/openssl.c | 3 ++-
+ 1 file changed, 2 insertions(+), 1 deletion(-)
+
+diff --git a/lib/vtls/openssl.c b/lib/vtls/openssl.c
+index 0b79fc50a9c5..4618beeb3867 100644
+--- a/lib/vtls/openssl.c
++++ b/lib/vtls/openssl.c
+@@ -1817,7 +1817,8 @@ CURLcode Curl_ossl_verifyhost(struct Curl_easy *data, struct connectdata *conn,
+               memcpy(peer_CN, ASN1_STRING_get0_data(tmp), peerlen);
+               peer_CN[peerlen] = '\0';
+             }
+-            result = CURLE_OUT_OF_MEMORY;
++            else
++              result = CURLE_OUT_OF_MEMORY;
+           }
+         }
+         else /* not a UTF8 name */

--- a/pkgs/tools/networking/curl/default.nix
+++ b/pkgs/tools/networking/curl/default.nix
@@ -66,6 +66,7 @@ stdenv.mkDerivation rec {
 
   patches = [
     ./7.79.1-darwin-no-systemconfiguration.patch
+    ./7.82.0-openssl-fix-CN-check.patch
   ];
 
   outputs = [ "bin" "dev" "out" "man" "devdoc" ];

